docs/madr/decisions/042-tcp-grpc-probe.md

```diff
@@ -0,0 +1,246 @@
+# TCP and gRPC probes
+
+- Status: 
+
+Technical Story: https://github.com/kumahq/kuma/issues/3875
+
+## Context and Problem Statement
+
+Kubernetes support three types of network based probes:
+- HTTP Get
+- TCP Socket
+- gRPC
+
+Kuma transparent proxy captures inbound traffic of pods, including the probe traffic. This leads to unexpected behaviour for Kubernetes probes without special handling: A TCP socket connection will always succeed to establish as it's redirected and handled by inbound passthrough listeners; HTTP requests can fail due to lack of mTLS client certificates attached, so are gRPC requests.
+
+We've implemented the [Virtual Probes feature](https://kuma.io/docs/2.6.x/policies/service-health-probes/#virtual-probes) which creates a dedicated listener for HTTP probes. And now, we want to support TCP socket and gRPC probes as well.
+
+Kubernetes users can specify HTTP, gRPC and TCP probes like this:
+
+```
+apiVersion: v1
+kind: Pod
+spec:
+  containers:
+  - name: app
+    startupProbe:
+      httpGet:
+        path: /healthz
+        port: 8080
+    livenessProbe:
+      grpc:
+        port: 2379
+        service: liveness
+    readinessProbe:
+      tcpSocket:
+        host: some-other-host
+        port: 1433
+```
+
+The `host` field in the `tcpSocket` probe is only used to resolve the network address when one wants to probe a different host than the pod's IP address and it's optional. The `service` field on gRPC probes is also optional. There is little space for us to carry contextual information on them.
+
+A pod can have multiple containers with each defines multiple and different probes, the worst case is a pod has all 3 different types of probes with many probes for each type.
```

Comment:
Isn't one of the main reason for these probes to control readiness of the sidecar? @michaelbeaumont what's the story of this with sidecar containers?

Reply:
Without a startupProbe, immediately after a sidecar container is started, the next initContainer is started. With a startupProbe, kubelet waits until it succeeds. Init containers otherwise can't have probes.

Reply:
I think the probes are to support the probes of the applications, rather than to probe the Kuma sidecar.
The kuma containers are run before the application container and probes of kuma containers are working as expected without any actions. So our virtual probes are to fix the probing features broken by the service mesh.